pkg/componentUpdater/componentUpdater.go

```diff
@@ -0,0 +1,41 @@
+package componentUpdater
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type UpdateFunction func(ctx context.Context, c client.Client, scheme *runtime.Scheme) error
```

Comment:
(nit) I think that instead of a function type, an interface might be more idiomatic

```golang
type Reconciler interface {
    Reconcile(context.Context, client.Client, *runtime.Scheme) error
}

// UpdateReconciler ensures that a resource exists with the desired state.
type UpdateReconciler struct {
    ms *stack.MonitoringStack
    resource client.Object
}

func (r *UpdateReconciler) Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error {
    if ms.Namespace == component.GetNamespace() {
    ...
    return nil
}

// DeleteReconciler ensures that a resource doesn't exist.
type DeleteReconciler struct {
    resource client.Object
}

func (r *DeleteReconciler) Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error {
    if err := c.Delete(ctx, r.resource); client.IgnoreNotFound(err) != nil {
    ...
}
```

We could have specialized reconcilers depending on the monitoring stack defintion:

```golang
// Executes a reconciler based on conditions.
type AlertmanagerReconciler struct {
   r Reconciler
}

func NewAlertmanagerReconciler(ms *stack.MonitoringStack) *AlertmanagerReconciler {
  var r Reconciler
  if ms.Spec.Alertmanager.Disabled {
    r = &DeleteReconciler{...}
  } else {
    r = &UpdateReconciler{...}
  }
  return &AlertmanagerReconciler{r: r}
}

func (r *AlertmanagerReconciler) Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error {
    return r.r.Reconcile(ctx, c, scheme)
}
```

Reply:
This is a good point. Even if an interface isn't strictly necessary (it has only one function) it is more idiomatic and can improve "testability".